code/code/misc/monster.cc

```diff
@@ -513,45 +513,37 @@ int TMonster::getMobDamage() const
   return max(1, idamrnd);
 }
 
-int TMonster::lookForEngaged(const TBeing *ch)
-{
-  // here's some AI
-  // tank and healer fight mob, tank flees, mob should now beat on
-  // healer (regardless of whether he engaged or not?)
-  if (isPc())
-    return FALSE;
+int TMonster::lookForEngaged(const TBeing *previousAttacker) {
+  if (isPc()) return false;
+
   // if someone is attacking me directly, bypass
-  if (attackers > 0)
-    return FALSE;
-  if (!isSmartMob(0))
-    return FALSE;
-
-  TThing * t;
-  for(StuffIter it=roomp->stuff.begin();it!=roomp->stuff.end();){
-    t=*(it++);
-    if (t == this || (ch && dynamic_cast<const TBeing *>(t) == ch))
-      continue;
-    TBeing *tbt = dynamic_cast<TBeing *>(t);
-    if (!tbt)
+  if (attackers > 0) return false;
+
+  for (auto thing : roomp->stuff) {
+    if (thing == this ||
+        (previousAttacker && dynamic_cast<const TBeing *>(thing) == previousAttacker))
       continue;
-    if (tbt->fight() == this) {
-      // I'm a smart mob, and, oh look, tbt is engaged with me...
-      int rc = takeFirstHit(*tbt);
+
+    auto thingAsTBeing = dynamic_cast<TBeing *>(thing);
+    if (!thingAsTBeing) continue;
+
+    if (thingAsTBeing->fight() == this) {
+      auto rc = takeFirstHit(*thingAsTBeing);
+
       if (IS_SET_DELETE(rc, DELETE_VICT)) {
-        delete tbt;
-        tbt = NULL;
+        delete thingAsTBeing;
```

Comment:
lookForEngaged is doing 2 things. Instead, it should just look through the room and find a mob attacking this, and return it. Then code somewhere else could hit that mob (possibly another method, or possibly just `doFlee`)

Reply:
After working on this some more, I realized that it was best for lookForEngaged to remain its own method in the TMonster class, because - though it's not coded that way now nor was it before - lookForEngaged _should_ also be called after a mob's opponent dies. 

Currently when that happens, if the mob has no active attackers when their opponent dies but there's someone engaged with them, the mob won't attack the engager but the engager will still be in combat.

I'm not familiar with the code that executes on mob/player death yet, so I'll try to address calling lookForEngaged there in a separate PR and just leave the on-death code unchanged for now.

As for the lookForEngaged code - I rewrote the function to use std::find_if with a lambda unary predicate to search for a potential opponent, which eliminated the need for two (nested) `for` loops. Everything still worked fine during my testing, so this should be a solid upgrade as far as syntax goes.

I also turned the code that handled the troglodyte racial perk into a lambda and was able to eliminate some code repetition.